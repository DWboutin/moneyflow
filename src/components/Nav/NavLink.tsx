import React, { SyntheticEvent } from "react";
import { Link } from "gatsby";
import styled from "styled-components";

import { ThemeContainer } from "../../styles/styles";
import { fontSizes } from "../../styles/typography";
import Boxicon, { BoxiconTypes } from "../BoxIcons/BoxIcons";

const Container = styled(Link)`
  display: flex;
  position: relative;
  flex-direction: row;
  height: 24px;
  align-items: center;
  padding-left: 60px;
  margin: 5px 0;
  color: ${({ theme }: ThemeContainer) => theme.nav.navigation.normal.text};
  text-decoration: none;
  font-size: ${fontSizes.normal};
  transition: all
    ${({ theme }: ThemeContainer) =>
      `${theme.animation.timing} ${theme.animation.easeInOutQuart}`};

  &::after {
    content: "";
    display: block;
    position: absolute;
    top: 0px;
    left: 0px;
    width: 0px;
    height: 100%;
    background-color: ${({ theme }: ThemeContainer) =>
      theme.nav.navigation.normal.icon};
    transition: all
      ${({ theme }: ThemeContainer) =>
        `${theme.animation.timing} ${theme.animation.easeInOutQuart}`};
  }

  & > .bx {
    margin-right: 10px;
    color: ${({ theme }: ThemeContainer) => theme.nav.navigation.normal.icon};
    transition: all
      ${({ theme }: ThemeContainer) =>
        `${theme.animation.timing} ${theme.animation.easeInOutQuart}`};
  }

  &:hover {
    color: ${({ theme }: ThemeContainer) => theme.nav.navigation.hover.text};

    &::after {
      width: 5px;
      background-color: ${({ theme }: ThemeContainer) =>
        theme.nav.navigation.hover.icon};
    }
    & > .bx {
      color: ${({ theme }: ThemeContainer) => theme.nav.navigation.hover.icon};
    }
  }

  &.active {
    color: ${({ theme }: ThemeContainer) => theme.nav.navigation.active.text};

    &::after {
      width: 5px;
      background-color: ${({ theme }: ThemeContainer) =>
        theme.nav.navigation.active.icon};
    }
    & > .bx {
      color: ${({ theme }: ThemeContainer) => theme.nav.navigation.active.icon};
    }
  }
`;

export interface Props {
  to: string;
  icon: string;
  iconType?: BoxiconTypes;
  label: string;
  onClick?: (e: SyntheticEvent<HTMLAnchorElement>) => void;
}

const NavLink: React.VoidFunctionComponent<Props> = ({
  to,
  icon,
  iconType,
  label,
  onClick,
}) => {
  return (
    <Container to={to} activeClassName="active" onClick={onClick}>
      <Boxicon name={icon} size="small" type={iconType} />
      <span>{label}</span>
    </Container>
  );
};

export default NavLink;
