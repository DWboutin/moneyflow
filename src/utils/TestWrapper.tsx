import React, { ReactNode } from "react";
import { Provider } from "react-redux";
import { ToastContainer } from "react-toastify";
import { ThemeProvider } from "styled-components";

import store from "../store/store";
import GlobalStyle from "../styles/globalStyle";

import { theme } from "../styles/styles";

interface Props {
  children: ReactNode;
}

const TestWrapper: React.FunctionComponent<Props> = ({ children }) => (
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <ToastContainer />
    </ThemeProvider>
  </Provider>
);

export default TestWrapper;
