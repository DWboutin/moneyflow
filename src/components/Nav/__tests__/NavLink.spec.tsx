import React from "react";
import { fireEvent, render, RenderResult } from "@testing-library/react";

import NavLink, { Props } from "../NavLink";
import TestWrapper from "../../../utils/TestWrapper";

describe("<NavLink />", () => {
  let renderResult: RenderResult;

  const mockOnClick = jest.fn();
  const renderNavLink = ({ to, icon, iconType, label, onClick }: Props) => {
    renderResult = render(
      <TestWrapper>
        <NavLink
          to={to}
          icon={icon}
          iconType={iconType}
          label={label}
          onClick={onClick}
        />
      </TestWrapper>
    );
  };

  describe("default rendering", () => {
    beforeEach(() => {
      renderNavLink({
        to: "/",
        icon: "dashboard",
        iconType: "solid",
        label: "Link",
        onClick: () => {},
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe("on click", () => {
    beforeEach(() => {
      renderNavLink({
        to: "/",
        icon: "dashboard",
        iconType: "solid",
        label: "Link",
        onClick: mockOnClick,
      });

      fireEvent(
        renderResult.container.firstChild as HTMLAnchorElement,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );
    });

    it("should have execute onClick handler", () => {
      expect(mockOnClick).toHaveBeenCalled();
    });
  });
});
