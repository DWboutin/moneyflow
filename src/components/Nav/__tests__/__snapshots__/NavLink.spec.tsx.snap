// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavLink /> default rendering should match snapshot 1`] = `
<div>
  <a
    class="sc-bdvvtL hEsoLv"
    href="/"
  >
    <i
      class="bx bxs-dashboard bx-sm"
    />
    <span>
      Link
    </span>
  </a>
  <div
    class="Toastify"
  />
</div>
`;
